src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +575,230 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
+
+	Oid schemaId = PG_GETARG_OID(0);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Ensure schema checks */
+	EnsureSchemaOwner(schemaId);
+
+	/* The schema should be a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (!IsTenantSchema(schemaId))
+	{
+		ereport(ERROR, (errmsg("schema %s is not a tenant schema", schemaName)));
+	}
+
+	ereport(NOTICE, (errmsg("converting the schema %s back to a regular schema",
+							schemaName)));
+
+	/* Only single shard tables are expected during the undistribution of the schema */
+	List *disallowedCitusTableTypesForUndistribute PG_USED_FOR_ASSERTS_ONLY =
+		DisallowedTableTypesForUndistributeSchema();
+	SchemaTables *schemaTables PG_USED_FOR_ASSERTS_ONLY = TablesUnderSchema(schemaId);
+	Assert(!SchemaHasCitusTableType(schemaTables,
+									disallowedCitusTableTypesForUndistribute));
+	Assert(schemaTables->postgresTableIdList == NIL);
+
+	/*
+	 * First, we need to delete schema metadata and sync it to workers. Otherwise,
+	 * we would get error from `ErrorIfTenantTable` while undistributing the tables.
+	 */
+	char *unregisterSchemaCommand = TenantSchemaDeleteCommand(schemaName);
+	DeleteTenantSchemaLocally(schemaId);
+	if (EnableMetadataSync)
+	{
+		SendCommandToWorkersWithMetadata(unregisterSchemaCommand);
+	}
+
+	/* Undistribute all single shard tables in the schema */
+	List *tablesToConvert = NIL;
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, schemaTables->singleShardTableIdList)
+	{
+		/*
+		 * Skip partitions as they would be undistributed by the parent table.
+		 *
+		 * We should filter out partitions here before undistributing the schema.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail with oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		/*
+		 * Error early before undistributing the schema if concurrent drop or some alter
+		 * statements occurred on the table. Note that we already had the AccessShareLock
+		 * lock when we call PartitionTable, so after this point we are guaranteed to
+		 * prevent concurrent drop and some modification on the table.
+		 */
+		EnsureRelationExists(relationId);
+		tablesToConvert = lappend_oid(tablesToConvert, relationId);
+	}
+
+	UndistributeTables(tablesToConvert);
+
+	/* Delete colocation metadata and sync to workers */
+	uint32 schemaColocationId = SchemaIdGetTenantColocationId(schemaId);
+	DeleteColocationGroup(schemaColocationId);
```

Comment:
Applies to line 816

```c
ereport(ERROR, (errmsg("%s is not allowed for %s because it is a tenant table"
```

becomes

```c
ereport(ERROR, (errmsg("%s is not allowed for %s because it belongs to a distributed schema"
```